### PR TITLE
Add documentation guide on writing a provider

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds a new documentation section on how to go about writing
a new provider. Right now there isn't really much documentation on the
procedure for writing a provider from scratch for a device or
simulator. While there is some api documentation on all the individual
pieces there was no guide putting everything together to explain the
higher level process. There used to be a tutorial on this, but that is a
bit stale at this point, and refers to the old interface. It also was
arguably in the wrong place since something like this should live in the
documentation with the interface so we can keep it up to date more
easily. This commit starts this guide and provides a high level overview
of all the components of a provider and how they need to be implemented.

### Details and comments

### CODE REVIEW

File: qiskit/providers/__init__.py [hunk 1]
```diff
@@ -110,6 +109,381 @@
...
+Writing a New Provider
+======================
+
+If you have a Quantum device or simulator that you would like to integrate with
+Qiskit the way to go about this is writing a provider. A provider will provide
+Terra with a method to get available :class:`~qiskit.provider.BackendV1` objects
+The :class:`~qiskit.provider.BackendV1` object provides both information
+describing a backend and its operation for the :mod:`~qiskit.transpiler` so that
+circuits can be compiled to something that is optimized and can execute on the
+backend and also the :meth:`~qiskit.providers.BackenvV1.run` method which can
```
Comment (nonhermitian) [inline qiskit/providers/__init__.py:122]: missing a "d" in backend

File: qiskit/providers/__init__.py [hunk 2]
```diff
@@ -110,6 +109,381 @@
...
+Qiskit Aqua, to get results from executing circuits on devices in a standard
+fashion regardless of how the Backend is implemented. For a simple example of
+a provider the
+`qiskit-aqt-provider <https://github.com/qiskit-community/qiskit-aqt-provider>`__
+packge is worth using as an example.
+
+Provider
+========
+
+A provider class serves a single purpose to get backend objects that enable
```
Comment (nonhermitian) [inline qiskit/providers/__init__.py:135]: I am not sure if we want to say providers only do backends, even though that is all we got now.  For example, there could be a transpiler pass manager that gets served up.

File: qiskit/providers/__init__.py [hunk 3]
```diff
@@ -110,6 +109,381 @@
...
+        def __init(self, provider):
+            super().__init__(
+                configuration=BackendConfiguration.from_dict(self.configuration),
+                provider=provider)
+
+        @classmethod
+        def _default_options(cls):
+            return Options(shots=1024, memory=False)
+
+        def run(circuits, **kwargs):
```
Comment (nonhermitian) [inline qiskit/providers/__init__.py:208]: Nice.

File: qiskit/providers/__init__.py [hunk 4]
```diff
@@ -110,6 +109,381 @@
...
+is only used if the backend either returns the necessary information.
+
+It's worth noting that this interface will likely evolve over time as the
+format, details, API, and information of the data the compiler needs from a
+backend will change as the transpiler and backends grow and improve. This is
+a key reason why the providers interfaces are versioned to enable making these
+changes in a controlled manner. The details here will change and reflect the
+latest version of this interface.
+
+Custom Basis Gates
```
Comment (nonhermitian) [inline qiskit/providers/__init__.py:246]: This is very good.  However, should this not be some place else like: https://qiskit.org/documentation/stubs/qiskit.circuit.EquivalenceLibrary.html ?  Perhaps we need some here and some there, not exactly sure.

**Review by nonhermitian — COMMENTED** [review]
Overall looks good, and adds some much needed material.  Just a few minor comments.

File: qiskit/providers/__init__.py [hunk 5]
```diff
@@ -110,6 +109,381 @@
...
+        def backends(self, name=None, **kwargs):
+            if name:
+                backends = [
+                    backend for backend in backends if backend.name() == name]
+            return filter_backends(backends, filters=filters, **kwargs)
+
+The key aspects to ensure are present is that any necessary information for
+authentication (if required) are present in the class and that the backends
+method matches the abstract interface and will return initialized Backend
+objects. The rest is up to the specific provider on how to implement.
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:163]: Big run on sentence in current form, also fairly awkward.  Needs punctuation.  Maybe use a bulleted list?

File: qiskit/providers/__init__.py [hunk 6]
```diff
@@ -110,6 +109,381 @@
...
+
+Backend
+=======
+
+The backend classes are the core to the provider. These class definitions
+define the interface between Qiskit and the hardware or simulator that will
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:169]: "These class definitions define" is bad.  Do you mean to say that the classes define the interface, or that they satisfy the interface etc etc?

File: qiskit/providers/__init__.py [hunk 7]
```diff
@@ -110,6 +109,381 @@
...
+execute circuits. It also defines the interface for a backend to provide the
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:170]:
```suggestion
execute circuits. They also define the interface for a backend to provide the
```

File: qiskit/providers/__init__.py [hunk 8]
```diff
@@ -110,6 +109,381 @@
...
+necessary information to describe the backend to the compiler so that it can
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:171]:
```suggestion
necessary information describing the backend to the compiler so that it can
```

File: qiskit/providers/__init__.py [hunk 9]
```diff
@@ -110,6 +109,381 @@
...
+embed and optimize any circuit to the backend. There are 3 required things
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:172]: optimize any circuit to the backend sounds weird.  Perhaps you mean "optimize any circuit on the backend" or something like that?

File: qiskit/providers/__init__.py [hunk 10]
```diff
@@ -110,6 +109,381 @@
...
+in every backend object, a backend configuration object set in the
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:173]:
```suggestion
in every backend object: a backend configuration object set in the
```

File: qiskit/providers/__init__.py [hunk 11]
```diff
@@ -110,6 +109,381 @@
...
+`__init__()`, a ``run()`` method, and a ``_default_options()`` method. For
+example, a minimum working example would be something like::
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:175]:
```suggestion
`__init__()` method, a ``run()`` method, and a ``_default_options()`` method. 
For example, a minimum working example would be something like::
```

File: qiskit/providers/__init__.py [hunk 12]
```diff
@@ -110,6 +109,381 @@
...
+
+Transpiler Interface
+--------------------
+
+The key piece of the backend object is how it describes itself to the compiler.
+This is composed of three classes, the
+:class:`~qiskit.providers.models.BackendConfiguration` which describes the
+immutable properties of the backend (things like the number of qubits, coupling
+map, etc), the :class:`~qiskit.providers.models.BackendProperties` which
+described measured properties of the device (such as current error rates), and
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:232]:
```suggestion
describes measured properties of the device (such as current error rates), and
```

File: qiskit/providers/__init__.py [hunk 13]
```diff
@@ -110,6 +109,381 @@
...
+the :class:`~qiskit.providers.models.PulseDefaults` which defines the
+default pulse behavior for the backend (such as the pulse sequences for gates).
+Of these the only required piece for a backend is the
+:class:`~qiskit.providers.models.BackendConfiguration`, the other information
+is only used if the backend either returns the necessary information.
```
Comment (elfrocampeador) [inline qiskit/providers/__init__.py:237]: This doesn't make sense.  Do you mean that the other information is only used if the backend _fails_ to return the necessary information?  Also, since you've been describing the backend classes, you can say "Of these only :class:`~qiskit.providers.models.BackendConfiguration` is required" or something like that.  Same meaning, but more concise.

**Review by elfrocampeador — COMMENTED** [review]
Copy editing and linguistic nitpicking

**Comment by jyu00** [comment]
The content looks good in general. But perhaps a bulleted list tldr in the beginning would make the task looks less daunting. Something like:
```
- Implement a `Provider` subclass that handles access to the backend(s).
- Implement a `Backend` subclass and its `run()` method. The class needs to describe the backend's configuration and options.
- Define basis gates for your backend(s).
- Implement a Job subclass with methods like `status()` and `result()`.
```

File: qiskit/providers/__init__.py (hunk 4 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py:246]: Yeah, I wasn't sure exactly where to put it. I settled on here because this has come up a couple of times in the context of writing a provider from scratch for a backend with custom gate definitions. I figured it was good here to provide concrete instructions on how to do this and how to ensure `transpile()` and `execute()` will work for compiling to a custom basis set.

**Comment by mtreinish** [comment]
> The content looks good in general. But perhaps a bulleted list tldr in the beginning would make the task looks less daunting. Something like:
> 
> ```
> - Implement a `Provider` subclass that handles access to the backend(s).
> - Implement a `Backend` subclass and its `run()` method. The class needs to describe the backend's configuration and options.
> - Define basis gates for your backend(s).
> - Implement a Job subclass with methods like `status()` and `result()`.
> ```

That's a good idea, I added this to the intro section in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/bfe06ad3092c8ade9e9bb02d2324b29d9150fbb0

File: qiskit/providers/__init__.py [hunk 14]
```diff
@@ -110,6 +109,394 @@
...
+                status = JobStatus.ERROR
+            return status
+
+    def submit(self):
+        raise NotImplementedError
+
+and for a sync job::
+
+    class MySyncJob(Job):
+        _async = True
```
Comment (mtreinish) [inline qiskit/providers/__init__.py:485]:
```suggestion
        _async = False
```

File: qiskit/providers/__init__.py [hunk 15]
```diff
@@ -110,6 +109,394 @@
...
+in every backend object: a backend configuration object set in the
+`__init__()` method, a ``run()`` method, and a ``_default_options()`` method.
+For example, a minimum working example would be something like::
+
+    from qiskit.providers import BackendV1 as Backend
+    from qiskit.providers.models import BackendConfiguration
+    from qiskit.providers import Options
+
+
+    class MVbackend(Backend):
```
Comment (lazyoracle) [inline qiskit/providers/__init__.py:195]: Typo `MVBackend` -> `MyBackend`

File: qiskit/providers/__init__.py [hunk 16]
```diff
@@ -110,6 +109,394 @@
...
+
+        configuration = {
+            'backend_name': 'my_backend',
+            'backend_version': '1.0.0',
+            'n_qubits': 5,
+            'basis_gates': ['p', 'sx', 'u', 'cx', 'id'],
+            'coupling_map': [[0, 1], [1, 2], [2, 3], [3. 4]],
```
Comment (lazyoracle) [inline qiskit/providers/__init__.py:202]: Typo `[3. 4]]` -> `[3, 4]]`

File: qiskit/providers/__init__.py [hunk 17]
```diff
@@ -110,6 +109,394 @@
...
+
+        def run(circuits, **kwargs):
+            # serialize circuits submit to backend and create a job
+            for kwarg in kwargs:
+                if not hasattr(kwarg, self.options):
+                    warnings.warn(
+                        "Option %s is not used by this backend" % kwarg,
+                        UserWarning, stacklevel=2)
+            options = {
+                'shots': kwargs.get('shots', self.options.shots)
```
Comment (lazyoracle) [inline qiskit/providers/__init__.py:229]: Typo missing `,` at the end of `'shots': kwargs.get('shots', self.options.shots)`

**Review by lazyoracle — CHANGES_REQUESTED** [review]

**Comment by lazyoracle** [comment]
I was trying to follow the steps outlined in this Docs Update to add my own provider, and I wanted to point out that the way `ProviderV1` is implemented, following these steps might not lead to a working 3rd-party provider. The example Provider (the one developed by AQT) linked in the docs in fact does not inherit the `ProviderV1` class provides its own implementation of the required methods as shown [here](https://github.com/qiskit-community/qiskit-aqt-provider/blob/de554d033cc949058906a6da0d472460d2f08c7d/qiskit_aqt_provider/aqt_provider.py#L21)

When you try to inherit the `ProviderV1` class and provide your implementation of `backends()`, you run into errors with conflicting names for attributes and methods. Its possible that I am just doing this wrong, so I tried to look for tests that check the `ProviderV1` implementation in #5086 which originally added this feature, but I didn't find any. 

Are there any examples of Providers (Internal or 3rd Party) that have been implemented making complete use of the steps outlined in these docs (i.e, inheriting the `ProviderV1` class)?

**Comment by mtreinish** [comment]
@lazyoracle yeah, the aqt provider was recently updated to enable tab completion for the `backends` attribute on it's `Provider` class so it no longer is using the `ProviderV1` abstract class as it's parent. The changes that were made there are something we're going to add into the `ProviderV2` abstract class. I still think it's a good example just because it's the simplest out there and it has less baggage than other providers. If you'd like to see it when it was using the `ProviderV1`  class you can see that in original migration commit: https://github.com/qiskit-community/qiskit-aqt-provider/commit/1f99e3f5cb30de8db7b6257b1ca1e7a44d7f15b3

That being said #5128 is also a good example you can refer to also. It's just a local simulator and returns sync job objects which is a bit different from other providers out there.

File: qiskit/providers/__init__.py [hunk 18]
```diff
@@ -110,6 +109,394 @@
...
+
+For a simple example of a provider the
+`qiskit-aqt-provider <https://github.com/qiskit-community/qiskit-aqt-provider>`__
+package is worth using as an example.
+
+Provider
+========
+
+A provider class serves a single purpose: to get backend objects that enable
+executing circuits on a device or simulator. The expecation is that any
```
Comment (taalexander) [inline qiskit/providers/__init__.py:149]:
```suggestion
executing circuits on a device or simulator. The expectation is that any
```

File: qiskit/providers/__init__.py [hunk 19]
```diff
@@ -110,6 +109,394 @@
...
+Writing a New Provider
+======================
+
+If you have a quantum device or simulator that you would like to integrate with
+Qiskit you will need to write a provider. A provider will provide Terra with a
+method to get available :class:`~qiskit.provider.BackendV1` objects. The
+:class:`~qiskit.provider.BackendV1` object provides both information describing
+a backend and its operation for the :mod:`~qiskit.transpiler` so that circuits
+can be compiled to something that is optimized and can execute on the
+backend and also the :meth:`~qiskit.providers.BackendV1.run` method which can
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:122]: This sentence became hard to follow, especially since it says both but then continues onto a third thing
```suggestion
backend. It also provides the :meth:`~qiskit.providers.BackendV1.run` method which can
```

File: qiskit/providers/__init__.py [hunk 20]
```diff
@@ -110,6 +109,394 @@
...
    JobV1
 
+======================
+Writing a New Provider
+======================
+
+If you have a quantum device or simulator that you would like to integrate with
+Qiskit you will need to write a provider. A provider will provide Terra with a
+method to get available :class:`~qiskit.provider.BackendV1` objects. The
+:class:`~qiskit.provider.BackendV1` object provides both information describing
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:119]: It might be a good idea to make this into a list instead. It could help writers of providers understand what their set of responsibilities are

Comment (lcapelluto) [reply qiskit/providers/__init__.py:119]: Hmm, I just read to the list below which I think is super clear. So maybe keep this the same format. I suggested below to break up the sentences

File: qiskit/providers/__init__.py [hunk 21]
```diff
@@ -110,6 +109,394 @@
...
+            self.token = token
+            self.backends = [MyBackend(provider=self)]
+
+        def backends(self, name=None, **kwargs):
+            if name:
+                backends = [
+                    backend for backend in backends if backend.name() == name]
+            return filter_backends(backends, filters=filters, **kwargs)
+
+The key aspects to ensure are present is that any necessary information for
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:173]: What about simply:
```suggestion
Ensure that any necessary information for
```

File: qiskit/providers/__init__.py [hunk 22]
```diff
@@ -110,6 +109,394 @@
...
+authentication (if required) are present in the class and that the backends
+method matches the abstract interface and will return initialized Backend
+objects. The rest is up to the specific provider on how to implement.
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:176]:
```suggestion
authentication (if required) are present in the class and that the backends
method matches the required interface. The rest is up to the specific provider on how to implement.
```
details about the backends method (returning initialized objects) could be written as a docstring in the mini class example

File: qiskit/providers/__init__.py [hunk 23]
```diff
@@ -110,6 +109,394 @@
...
+Backend
+=======
+
+The backend classes are the core to the provider. These class definitions
+define the interface between Qiskit and the hardware or simulator that will
+execute circuits. They also define the interface for a backend to provide the
+necessary information describing the backend to the compiler so that it can
+embed and optimize any circuit to the backend. There are 3 required things
+in every backend object: a backend configuration object set in the
+`__init__()` method, a ``run()`` method, and a ``_default_options()`` method.
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:187]:
```suggestion
in every backend object: a backend configuration property,
a ``run()`` method, and a ``_default_options()`` method.
```

File: qiskit/providers/__init__.py [hunk 24]
```diff
@@ -110,6 +109,394 @@
...
+describes measured properties of the device (such as current error rates), and
+the :class:`~qiskit.providers.models.PulseDefaults` which defines the
+default pulse behavior for the backend (such as the pulse sequences for gates).
+Of these the only required piece for a backend is the
+:class:`~qiskit.providers.models.BackendConfiguration`, the other information
+is only used if the backend either returns the necessary information.
+
+It's worth noting that this interface will likely evolve over time as the
+format, details, API, and information of the data the compiler needs from a
+backend will change as the transpiler and backends grow and improve. This is
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:254]:
```suggestion
It's worth noting that this interface will likely evolve over time as the
compiler grows and improves, thus requiring new information from the backends. This is
```

File: qiskit/providers/__init__.py [hunk 25]
```diff
@@ -110,6 +109,394 @@
...
+
+        def _define(self):
+            qc = QuantumCircuit(1)
+            q.ry(np.pi / 2, 0)
+            self.definition = qc
+
+The key thing to ensure that any custom gates in your Backend's basis set
+is that your custom gate's name attribute (the first param on
+``super().__init__()`` in the ``__init__`` defintion above does not conflict
+with the name of any other gates. The name attribute is what is used to
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:285]: re read this. missing paren; seems like there's a typo in the first phrase

File: qiskit/providers/__init__.py [hunk 26]
```diff
@@ -110,6 +109,394 @@
...
+identify the gate in the basis set for the transpiler and if there is a conflict
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:286]:
```suggestion
identify the gate in the basis set for the transpiler. If there is a conflict
```

File: qiskit/providers/__init__.py [hunk 27]
```diff
@@ -110,6 +109,394 @@
...
+the transpiler will not know which gate to use.
+
+After you've defined the custom gates to use for the Backend's basis set
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:289]: Can these steps be numbered?

File: qiskit/providers/__init__.py [hunk 28]
```diff
@@ -110,6 +109,394 @@
...
+then you need to add equivalence rules to the standard equivalence library
+so that the :class:`~qiskit.transpiler.passes.BasisTranslator` used in the
+:func:`~qiskit.compiler.transpile` function and :mod:`~qiskit.transpiler`
+module can convert an arbitrary circuit using the custom basis set. This
+can be done by defining equivalent circuits, in terms of the custom gate,
+for standard gates. Typically if you can convert from a
+:class:`~qiskit.circuit.library.CXGate` (if your basis doesn't include a
+standard 2 qubit gate) and that some commonly used single
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:297]:
```suggestion
standard 2 qubit gate) and some commonly used single
```

File: qiskit/providers/__init__.py [hunk 29]
```diff
@@ -110,6 +109,394 @@
...
+identify the gate in the basis set for the transpiler and if there is a conflict
+the transpiler will not know which gate to use.
+
+After you've defined the custom gates to use for the Backend's basis set
+then you need to add equivalence rules to the standard equivalence library
+so that the :class:`~qiskit.transpiler.passes.BasisTranslator` used in the
+:func:`~qiskit.compiler.transpile` function and :mod:`~qiskit.transpiler`
+module can convert an arbitrary circuit using the custom basis set. This
+can be done by defining equivalent circuits, in terms of the custom gate,
+for standard gates. Typically if you can convert from a
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:295]: if you can covert from these, then what?

File: qiskit/providers/__init__.py [hunk 30]
```diff
@@ -110,6 +109,394 @@
...
+The key thing to ensure that any custom gates in your Backend's basis set
+is that your custom gate's name attribute (the first param on
+``super().__init__()`` in the ``__init__`` defintion above does not conflict
+with the name of any other gates. The name attribute is what is used to
+identify the gate in the basis set for the transpiler and if there is a conflict
+the transpiler will not know which gate to use.
+
+After you've defined the custom gates to use for the Backend's basis set
+then you need to add equivalence rules to the standard equivalence library
+so that the :class:`~qiskit.transpiler.passes.BasisTranslator` used in the
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:291]: mentioning BasisTranslator might be too much info. if you refactor here, remove from below as well. I think the fact that the transpiler needs this info is probably enough

File: qiskit/providers/__init__.py [hunk 31]
```diff
@@ -110,6 +109,394 @@
...
+    SessionEquivalenceLibrary.add_equivalence(
+        U2Gate(phi, lam), def_sy_u2)
+
+
+You will want this to be run on import so that as soon as the provider's
+package is imported it will be run. This will ensure that any time the
+:class:`~qiskit.transpiler.passes.BasisTranslator` pass is run with the
+custom gates the equivalence rules are defined.
+
+It's also worth noting that depending on the basis your using some optimization
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:339]:
```suggestion
It's also worth noting that depending on the basis you're using, some optimization
```

File: qiskit/providers/__init__.py [hunk 32]
```diff
@@ -110,6 +109,394 @@
...
+passes in the transpiler, such as
+:class:`~qiskit.transpiler.passes.Optimize1qGatesDecomposition`, may not
+be able to operate with your custom basis. For example continuing with the
+``SYGate`` example the
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:343]:
```suggestion
be able to operate with your custom basis. For  our``SYGate`` example, the
```

File: qiskit/providers/__init__.py [hunk 33]
```diff
@@ -110,6 +109,394 @@
...
+:class:`~qiskit.transpiler.passes.Optimize1qGatesDecomposition` will not be
+able to simplify runs of single qubit gates into the SY basis. This is because
+the :class:`~qiskit.quantum_info.OneQubitEulerDecomposer` class does not
+know how to work in the SY basis. To solve this the ``SYGate`` class would
+need to be added to Qiskit and
+:class:`~qiskit.quantum_info.OneQubitEulerDecomposer` updated to support
+decomposing to the ``SYGate``. Longer term that is likely a better direction
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:350]: Is this info documented in the transpiler? It would be nicer here to be able to say "Some things won't work for any basis, if you run into this problem, look at the transpiler module docs"

File: qiskit/providers/__init__.py [hunk 34]
```diff
@@ -110,6 +109,394 @@
...
+
+Of key importance is the :meth:`~qiskit.providers.BackendV1.run` method, which
+is used to actually submit circuits to a device or simulator. The run method
+handles submitting the circuits to the backend to be executed and returning a
+:class:`~qiskit.providers.Job` object. Depending on the type of backend this
+typically involves serializing the circuit object into the API format used by a
+backend. For example, on IBMQ backends from the ``qiskit-ibmq-provider``
+package this involves converting from a quantum circuit and options into a
+`qobj <https://arxiv.org/abs/1809.03452>`__ JSON payload and submitting
+that to the IBM Quantum API. Since every backends interface is different (and
```
Comment (lcapelluto) [inline qiskit/providers/__init__.py:366]:
```suggestion
that to the IBM Quantum API. Since every backend interface is different (and
```

**Review by lcapelluto — CHANGES_REQUESTED** [review]
👍 minor comments

File: qiskit/providers/__init__.py (hunk 33 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py:347]: It doesn't look like it, looking at the docs for the pass itself: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py there is not really any detail in there (which is my fault). 

The transpile documentation doesn't really go into this level of detail either:

- https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/__init__.py is pretty high level about the theory of operation
-  https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/compiler/transpiler.py is about the mechanics of the api for calling `transpile`

The one place that has this documented is: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/quantum_info/synthesis/one_qubit_decompose.py but you'd need to know that the `Optimize1qGatesDecomposition`  pass is using that internally.

File: qiskit/providers/__init__.py (hunk 29 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py:293]: This continues on the next line the full sentence reads   "Typically if you can convert from a `CXGate` (if your basis doesn't include a standard 2 qubit gate) and some commonly used single qubit rotation gates like `HGate`."

There is another issue here though, the next sentence also starts with "Typically", which I'll fix now

Comment (mtreinish) [reply qiskit/providers/__init__.py:293]: Oh, nm I see what you're saying there's no finish there. I'll fix that now

File: qiskit/providers/__init__.py (hunk 30 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Yep I agree, I removed it in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/922ae3cc737a960f608e297abdcda8d16f0fb0b0

File: qiskit/providers/__init__.py (hunk 29 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/922ae3cc737a960f608e297abdcda8d16f0fb0b0

File: qiskit/providers/__init__.py (hunk 27 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: There are only 2 steps (but they need to be repeated for each gate), but I added the numbers in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/922ae3cc737a960f608e297abdcda8d16f0fb0b0

File: qiskit/providers/__init__.py (hunk 25 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Yeah, it reads a bit awkwardly, I tried to clean it up in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/922ae3cc737a960f608e297abdcda8d16f0fb0b0

File: qiskit/providers/__init__.py (hunk 6 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/922ae3cc737a960f608e297abdcda8d16f0fb0b0

File: qiskit/providers/__init__.py (hunk 9 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: I went with "optimize any circuit for the backend" in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/922ae3cc737a960f608e297abdcda8d16f0fb0b0 is that better?

File: qiskit/providers/__init__.py (hunk 13 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/5406/commits/922ae3cc737a960f608e297abdcda8d16f0fb0b0 (I also dropped the second half about the optional pieces it was redundant and confusing as you pointed out.

File: qiskit/providers/__init__.py [hunk 35]
```diff
@@ -110,6 +109,391 @@
...
+            super().__init__("sy", 1, [], label=label)
+
+        def _define(self):
+            qc = QuantumCircuit(1)
+            q.ry(np.pi / 2, 0)
+            self.definition = qc
+
+The key thing to ensure is that any custom gates in your Backend's basis set
+is that your custom gate's name attribute (the first param on
+``super().__init__()`` in the ``__init__`` defintion above) does not conflict
```
Comment (kdk) [inline qiskit/providers/__init__.py:281]:
```suggestion
The key thing to ensure is that for any custom gates in your Backend's basis set
your custom gate's name attribute (the first param on
``super().__init__()`` in the ``__init__`` defintion above) does not conflict
```

File: qiskit/providers/__init__.py [hunk 36]
```diff
@@ -110,6 +109,391 @@
...
+    from qiskit.circuit.library import U2Gate
+
+
+    # H => Z SY
+    q = qiskit.QuantumRegister(1, "q")
+    def_sy_h = qiskit.QuantumCircuit(q)
+    def_sy_h.append(ZGate(), [q[0]], [])
+    def_sy_h.append(SYGate(), [q[0]], [])
+    SessionEquivalenceLibrary.add_equivalence(
+        HGate(), def_sy_h)
```
Comment (kdk) [inline qiskit/providers/__init__.py:318]: It might be worth mentioning that this can also be done through `HGate().add_decomposition(def_sy_h)` (though for sake of clarity, having the full example here makes sense).

File: qiskit/providers/__init__.py [hunk 37]
```diff
@@ -110,6 +109,391 @@
...
+
+You will want this to be run on import so that as soon as the provider's
+package is imported it will be run. This will ensure that any time the
+:class:`~qiskit.transpiler.passes.BasisTranslator` pass is run with the
+custom gates the equivalence rules are defined.
+
+It's also worth noting that depending on the basis you're using, some optimization
+passes in the transpiler, such as
+:class:`~qiskit.transpiler.passes.Optimize1qGatesDecomposition`, may not
+be able to operate with your custom basis. For  our``SYGate`` example, the
```
Comment (kdk) [inline qiskit/providers/__init__.py:340]:
```suggestion
be able to operate with your custom basis. For our ``SYGate`` example, the
```

**Review by kdk — APPROVED** [review]
Thanks @mtreinish .
